File: dynamic_listings/javascript/default/design_patterns/singleton.js
class LazySingleton {

    // This way the constructor always returns the singleton instance
    constructor(){
        // Multi-threading: manage race conditions
        // ----- Critical region start -----
        if (!Singleton.INSTANCE){
            Singleton.INSTANCE = this;
        }
        // ----- Critical region end -----
        return Singleton.INSTANCE;
    }

    static getInstance(){
        return Singleton.INSTANCE;
    }
}

File: dynamic_listings/rust/default/design_patterns/singleton.rs
struct LazySingleton;

impl LazySingleton {
    fn new() -> Self {
        Self
    }

    pub fn instance() -> &'static Self {
        static INSTANCE: std::sync::LazyLock<LazySingleton> =
            std::sync::LazyLock::new(LazySingleton::new);

        &*INSTANCE
    }
}
